source/manual/intro-to-docker.html.md

```diff
@@ -0,0 +1,317 @@
+
+* Create a persistent volume for our gems:
+
+```shell
+$mac docker volume create content-publisher-bundle
+```
+
+*   Run docker with that volume by using the `-v` flag and passing it the name of the volume:
+
+```shell
+$mac docker run -it --rm -v $(pwd):/app -v `content-publisher-bundle`:/usr/local/bundle ruby:2.6.3 bash
+```
+
+*   Then install the gems again:
+
+```shell
+$dev cd /app
+$dev bundle install
+```
+
+This time the gems will install on the `content-publisher-bundle` which is mapped to the container at the `/usr/local/bundle` path.
+
+### Specifying dependencies (node, PostgreSQL, Chrome)
+
+Now we have our gems installed, we can try and run the content-publisher tests:
+
+```shell
+$dev cd /app
+$dev bundle exec rake
+```
+
+We get an error: “could not find javascript runtime” - we need Node installed on our container. We can follow standard commands to install Node, but this won't persist and we'll have the same problem that we had above with installing gems.
+
+It seems like our Ruby image isn’t doing quite what we need, so we’re going to create our own image based off it.
+
+#### Create ruby + node image
+
+ We are going to create our own Docker image based off of the ruby one, and add in other dependencies such as node.
+
+To create our own image, we need a Dockerfile. A Dockerfile normally starts with a `FROM [image]` which bases your new image off an existing image. We can then execute other commands by prefixing them with `RUN`. Each RUN command adds a new layer. You can think of it as each step creating a new image, but we only care about the final image to come out of the last step.
+Create a Dockerfile in the content-publisher project:
+
+> Note: there will likely already be a Dockerfile, so just remove it for the tutorial.
+
+*   Create a `Dockerfile` in the root of the [content-publisher] project:
+
+```docker
+FROM ruby:2.6.3
+
+RUN curl -sL https://deb.nodesource.com/setup_12.x | bash
+
+RUN apt-get install nodejs && apt-get clean
+```
+
+*   Build our new image:
+
+```shell
+$mac docker build -t content-publisher
+```
+
+This creates an image based on the Dockerfile we just added.
+
+`-t` tags, or names it, as `content-publisher`.
+
+*   Now we can start a container with our new image:
+
+```shell
+$mac docker run -it --rm -v $(pwd):/app -v content-publisher-bundle:/usr/local/bundle content-publisher bash
+```
+
+Each time you change the Dockerfile, you need to remember to rebuild the docker image.
+
+If we run the tests again, this time we see another problem about no database. That's because we haven't got PostgreSQL installed.
+
+#### Use PostgreSQL image/container
+
+Our Javascript errors are now fixed, but we're getting new errors because we don't have a database. Docker containers are quick to start and it's possible for them to talk to each other, so let's set up a separate container for postgres.
```

Comment:
The first sentence here repeats the sentence immediately above the last heading - I think we can remove one of them.